### ISSUE TITLE
Tuple fields are silently ignored

### ISSUE THREAD

**Issue by mqudsi** [aspnet/EntityFrameworkCore#14661]
Given a `DbSet<T>` where `T` contains a publicly-visible, non-ignored tuple property, it seems that EF Core completely ignores it at the time of creation rather than mapping it somehow or else throwing an exception about an unsupported type.

```csharp
    public class DriverLocation
    {
        public int Id { get; set; }
        public DateTimeOffset Timestamp { get; set; }
        public int DriverId { get; set; }
        [Required]
        [ForeignKey("DriverId")]
        public int Driver { get; set; }
        public (float, float) Position;
    }

    public class ApplicationDbContext : IdentityDbContext<ApplicationUser>
    {
        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
            : base(options)
        {
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);
        }

        public DbSet<DriverLocation> DriverLocationLog { get; set; }
    }
```

this gets mapped to 

```csharp
            migrationBuilder.CreateTable(
                name: "DriverLocationLog",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("Sqlite:Autoincrement", true),
                    Timestamp = table.Column<DateTimeOffset>(nullable: false),
                    DriverId = table.Column<int>(nullable: false),
                    Driver = table.Column<int>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_DriverLocationLog", x => x.Id);
                });
```

with the `DriverLocationLog.Postion` field being dropped altogether.

My working assumption was that named or anonymous tuples would be expanded to named or anonymous columns, provided that the tuples are composed only of primitive/supported types.

### Further technical details
EF Core version: 2.2.1
Database Provider: SQLite
Operating system: Windows 10 17763
IDE: Visual Studio 2019 16.0.0 preview 2.2

**Comment by roji** [comment]
If/when implementing this, don't get in the way of providers which can map tuples natively (the idea is to map this to PostgreSQL composite types).

**Comment by ajcvickers** [comment]
Triage: putting this on the backlog. It's a nice idea that could work with a value converter convention. This would only kick in if the provider doesn't natively support some mapping, so PostgreSQL should be fine. However, it will require #13947 to be implemented first.

**Comment by dmitryshunkov** [comment]
@roji is there a way to do it with Npgsql currently, as EF Core doesn't support it yet? I have composite primary key `(Id, Type)` and want to generate next SQL:
```
SELECT *
FROM Table
WHERE (Id, Type) IN ((1, 1), (2, 1))
```
It is possible to generate IN clause for one column, using `Contains` method but I cannot figure out if it possible for columns within composite type. Maybe there is a workaround?

**Comment by roji** [comment]
@dmitryshunkov I don't think that's supported at the moment (maybe @smitpatel knows something I don't).

We could make this expressible with the following, assuming blog has a composite key:

```c#
var blogs = ctx.Blogs.Where(b => new[]
{
    new Blog { Id = 1, Type = 1 },
    new Blog { Id = 2, Type = 1 }
}.Contains(b)).ToList();
```

However, please open a separate issue as this is unrelated with the current issue.

**Comment by smitpatel** [comment]
@roji - Is that valid Sql in postgre? If yes then possibly provider need to override SqlTranslator and put tuple as translated SqlExpression as it can be generated as literal. (Relational one returns null atm)

**Comment by dmitryshunkov** [comment]
@roji, in my opinion, composite key shouldn't affect the possibility of building query with other columns in row constructor for subquery. Potential syntax could be something like this:
```
var blogs = ctx.Blogs
   .Where(p => (new[] { new { Id = 1, Type = 1, Title = "New" } })
      .Contains(new { p.Id, p.Type, p.Title }))
   .ToList();
```
to generate SQL:
```
SELECT *
FROM Blogs
WHERE (Id, Type, Title) IN ((1, 1, 'New'))
```

**Comment by roji** [comment]
@smitpatel yeah, the following works in PostgreSQL, I guess it's yet another provider-specific feature :)

```sql
roji=# SELECT 1 WHERE (1, 2) IN ((1, 2), (3, 4));
 ?column? 
----------
        1
(1 row)

roji=# SELECT 1 WHERE (1, 4) IN ((1, 2), (3, 4));
 ?column? 
----------
(0 rows)
```

Re the syntax, sure that looks good. My proposal above is the "entity equality" version which seems like it could be useful as well.

Opened https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/898 to track.

**Comment by marchy** [comment]
Just wanted to post an inter-dependency for when this gets taken on: namely that [nullable/optional tuples ought to be supported](https://github.com/dotnet/efcore/issues/31376) for this.

One of the core benefits of tuples is that they allow encapsulating multiple fields that belong together in a logical unit – thereby instead of modelling them as multiple independent nullable fields in the domain (ie: `Field1:T1?`, `Field2:T2?`), you can have a tuple that is nullable itself, but allows the fields to have nullability specified individually (ie: `(Field1:T1, Field2:T2)?`). For example: a price and its currency – should either neither be set or both be set, but never only one or the other.

Please allow nullable tuples to be modelled when this gets taken on.
